File: tauri-plugin-device/android/src/main/java/com/plugin/device/handlers/DeviceInfoHandler.kt
package com.plugin.device.handlers

import android.app.Activity
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.BatteryManager
import android.os.Build
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject

class DeviceInfoHandler(private val activity: Activity) {
    
    fun getDeviceInfo(invoke: Invoke) {
        val ret = JSObject()
        ret.put("osName", "Android")
        ret.put("osVersion", Build.VERSION.RELEASE)
        ret.put("deviceModel", Build.MODEL)
        ret.put("deviceManufacturer", Build.MANUFACTURER)
        invoke.resolve(ret)
    }
    
    fun getNetworkInfo(invoke: Invoke) {
        val ret = JSObject()
        
        val connectivityManager = activity.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = connectivityManager.activeNetwork
        
        if (network == null) {
            ret.put("isConnected", false)
            ret.put("networkType", "none")
            invoke.resolve(ret)
            return
        }
        
        ret.put("isConnected", true)
        
        val capabilities = connectivityManager.getNetworkCapabilities(network)
        val networkType = when {
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true -> "wifi"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) == true -> "ethernet"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true -> "cellular"
            else -> "unknown"
        }
        
        ret.put("networkType", networkType)
        invoke.resolve(ret)
    }
    
    fun getBatteryInfo(invoke: Invoke) {
        val ret = JSObject()
        
        val batteryManager = activity.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        val batteryLevel = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        val isCharging = batteryManager.isCharging
        
        ret.put("level", batteryLevel)
        ret.put("isCharging", isCharging)
        
        invoke.resolve(ret)
    }
}

File: tauri-plugin-device/ios/Sources/DevicePlugin/DeviceInfoHandler.swift
import SwiftRs
import Tauri
import UIKit
import WebKit

#if os(macOS)
import AppKit
#endif

class DeviceInfoHandler: NSObject {
    
    func getDeviceInfo(_ invoke: Invoke) throws {
        #if os(iOS)
        let osName = "iOS"
        let osVersion = UIDevice.current.systemVersion
        let deviceModel = UIDevice.current.model
        #elseif os(macOS)
        let osName = "macOS"
        let version = ProcessInfo.processInfo.operatingSystemVersion
        let osVersion = "\(version.majorVersion).\(version.minorVersion).\(version.patchVersion)"
        let deviceModel = "Mac"
        #else
        let osName = "Unknown"
        let osVersion = "Unknown"
        let deviceModel = "Unknown"
        #endif
        
        invoke.resolve([
            "osName": osName,
            "osVersion": osVersion,
            "deviceModel": deviceModel,
            "deviceManufacturer": "Apple"
        ])
    }
    
    func getNetworkInfo(_ invoke: Invoke) throws {
        let networkStatus = getNetworkStatus()
        invoke.resolve(networkStatus)
    }
    
    func getBatteryInfo(_ invoke: Invoke) throws {
        UIDevice.current.isBatteryMonitoringEnabled = true
        
        let level: Int
        if UIDevice.current.batteryLevel < 0 {
            level = 100
        } else {
            level = Int(UIDevice.current.batteryLevel * 100)
        }
        
        let isCharging = UIDevice.current.batteryState == .charging || 
                         UIDevice.current.batteryState == .full
        
        invoke.resolve([
            "level": level,
            "isCharging": isCharging
        ])
    }
    
    private func getNetworkStatus() -> [String: Any] {
        return [
            "isConnected": true,
            "networkType": "unknown"
        ]
    }
}
